### PR TITLE
Fixing incorrectly assigned goal/plan pairs

### PR DESCRIPTION
**Pull request by ayrton04**
@jasonimercer @p6chen @skaynama

### CODE REVIEW

File: move_base/src/move_base.cpp [hunk 1]
```diff
@@ -647,13 +648,27 @@ namespace move_base {
...
+
+      // makePlan can take time. While planning, it's possible that we received a new goal. We should check for this
+      // condition, and if it's true, we should carry on as if we did not succeed in planning this goal, which will
+      // cause us to move on and re-start this thread for the new goal.
+      lock.lock();
+      geometry_msgs::PoseStamped latest_goal = planner_goal_;
+      lock.unlock();
+
+      // Make sure the goal didn't change
+      if (distanceXYTheta(latest_goal, plan_goal) >= 1e-3)
```
Comment (ayrton04) [inline move_base/src/move_base.cpp]: I am checking for a changed goal while we planned. I could instead check the end pose in the plan further down, but that I'm concerned that it would make assumptions about how the planner works (e.g., there may be a planner that doesn't carry us all the way to the goal intentionally).

File: move_base/src/move_base.cpp [hunk 2]
```diff
@@ -647,13 +648,27 @@ namespace move_base {
...
+      {
+        ROS_INFO_STREAM("Planned goal and latest received goal do not match. Assuming new goal issued; ignoring this plan.\n");
```
Comment (skaynama) [inline move_base/src/move_base.cpp]: Would a warn make more sense here than an info?

File: move_base/src/move_base.cpp [hunk 3]
```diff
@@ -647,13 +648,27 @@ namespace move_base {
...
+        gotPlan = false;
```
Comment (skaynama) [inline move_base/src/move_base.cpp]: We should probably reset ```last_valid_plan_``` timer as well so that we don't trigger recoveries unnecessarily and just cycle through the next planning iteration.

File: move_base/src/move_base.cpp (hunk 2 above)
Comment (ayrton04) [inline move_base/src/move_base.cpp]: I'll defer to you. As I see it, having goals interrupted is part of the normal operation of the robot, and we're just stating that we detected the event. To me, a warning says that we need to be concerned about something, or that something wasn't as it should have been.

File: move_base/src/move_base.cpp (hunk 3 above)
Comment (ayrton04) [inline move_base/src/move_base.cpp]: Ok, will do.

File: move_base/src/move_base.cpp (hunk 2 above)
Comment (skaynama) [inline move_base/src/move_base.cpp]: I have no strong preference on this. I see your point. Leave it as is.

**Comment by skaynama** [comment]
Other than the minor comment on the timer, LGTM.

**Comment by p6chen** [comment]
lgtm

**Comment by jasonimercer** [comment]
LGTM

File: move_base/src/move_base.cpp [hunk 4]
```diff
@@ -647,13 +648,27 @@ namespace move_base {
...
       planner_plan_->clear();
-      bool gotPlan = n.ok() && makePlan(temp_goal, *planner_plan_);
+      bool gotPlan = n.ok() && makePlan(plan_goal, *planner_plan_);
+
+      // makePlan can take time. While planning, it's possible that we received a new goal. We should check for this
+      // condition, and if it's true, we should carry on as if we did not succeed in planning this goal, which will
+      // cause us to move on and re-start this thread for the new goal.
+      lock.lock();
+      geometry_msgs::PoseStamped latest_goal = planner_goal_;
+      lock.unlock();
```
Comment (servos) [inline move_base/src/move_base.cpp]: I would need to check closer but I think there is still a very small chance of a race condition here. The plan could change between here and when you set planner_plan_ = latest_plan_ (line 679). To ensure it is doesn't you probably need to move this logic into the same lock block with that planner_plan_ set.

Comment (ayrton04) [inline move_base/src/move_base.cpp]: Good call. I'll wrap this in with @skaynama's change.

**Comment by ayrton04** [comment]
Would you mind giving it a second look, @skaynama and @servos? Both of your suggested changes were related.

File: move_base/src/move_base.cpp [hunk 5]
```diff
@@ -833,7 +851,9 @@ namespace move_base {
 
       //if we're done, then we'll return from execute
       if(done)
+      {
         return;
+      }
```
Comment (skaynama) [inline move_base/src/move_base.cpp]: IMO we should minimize the amount of unnecessary change in navigation as we're aiming to sync with upstream soon and these will make it difficult to figure out where the real divergences are. Ditto for ```distance``` vs. ```distanceXY``` below.

**Comment by skaynama** [comment]
LGTM.

File: move_base/src/move_base.cpp (hunk 5 above)
Comment (ayrton04) [inline move_base/src/move_base.cpp]: OK. Want me to pull these changes back out? I can just add my new distanceXYTheta method and keep the orginal distance method as-is.

**Comment by servos** [comment]
LTGM
